### PR TITLE
Add a plugin for making curl-like requests with various protocols (only http/https for now)

### PR DESCRIPTION
**Pull request by zeroSteiner**
This PR introduces a requests plugin which supplies the `request` command. The purpose is to allow making various protocol requests from within metasploit using a curl-like syntax that many people are familiar with. Only http and https are supported at this time, but the code was changed to make it easy to add new protocols by simply defining parse_args_newproto and a handle_request_newproto methods.  The plugin will automatically start recognizing the new protocol as being supported (tho the 'newproto' must match the URI handler, so to handle ftp:// the methods should be parse_args_ftp and handle_request_ftp.  The http/https implementation here uses the Rex HTTP client and thus supports the internal routing provided by metasploit. Some times users need to make a quick HTTP request over a pivoted host, this should help make that a lot easier.  And, adding new protocols should be easy.

There is currently support for the most commonly used flags from curl.

Help menu showing all the flags and options:

```
msf> load request
[*] Successfully loaded plugin: Request
msf > request
Usage: request [options] uri
Supported uri types are: https://, http://
To see usage for a specific uri type, use request -h uri
msf > request -h http://
Usage: request [options] uri

OPTIONS:

    -0        Use HTTP 1.0
    -1        Use TLSv1 (SSL)
    -2        Use SSLv2 (SSL)
    -3        Use SSLv3 (SSL)
    -A <opt>  User-Agent to send to server
    -G        Send the -d data with an HTTP GET
    -H <opt>  Custom header to pass to server
    -I        Show document info only
    -X <opt>  Request method to use
    -d <opt>  HTTP POST data
    -h        This help text
    -i        Include headers in the output
    -o <opt>  Write output to <file> instead of stdout
    -u <opt>  Server user and password

```

Examples (using the [Github API](https://developer.github.com/v3/) as a target endpoint):

Showing headers:

```
msf> request -I https://api.github.com/users/octocat/orgs
HTTP/1.1 200 OK
Server: GitHub.com
Date: Fri, 13 Mar 2015 01:10:03 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 2
Status: 200 OK
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 55
X-RateLimit-Reset: 1426209515
Cache-Control: public, max-age=60, s-maxage=60
ETag: "d751713988987e9331980363e24189ce"
Vary: Accept, Accept-Encoding
X-GitHub-Media-Type: github.v3; format=json
X-XSS-Protection: 1; mode=block
X-Frame-Options: deny
Content-Security-Policy: default-src 'none'
Access-Control-Allow-Credentials: true
Access-Control-Expose-Headers: ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval
Access-Control-Allow-Origin: *
X-GitHub-Request-Id: 4CBC45A9:56A1:114821:550238EB
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Served-By: 8a5c38021a5cd7cef7b8f49a296fee40
```

Submitting credentials via basic authentication:

```
msf> request https://api.github.com -u foo:bar
{"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}
```

Connecting with specific SSL versions:

```
msf> request -2 https://api.github.com/
[-] Encountered an SSL error
msf>request -3 https://api.github.com/
[-] Encountered an SSL error
msf> request -1 https://api.github.com/
{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","notifications_url":"https://api.github.com/notifications","organization_repositories_url":"https://api.github.com/orgs/{org}/repos{?type,page,per_page,sort}","organization_url":"https://api.github.com/orgs/{org}","public_gists_url":"https://api.github.com/gists/public","rate_limit_url":"https://api.github.com/rate_limit","repository_url":"https://api.github.com/repos/{owner}/{repo}","repository_search_url":"https://api.github.com/search/repositories?q={query}{&page,per_page,sort,order}","current_user_repositories_url":"https://api.github.com/user/repos{?type,page,per_page,sort}","starred_url":"https://api.github.com/user/starred{/owner}{/repo}","starred_gists_url":"https://api.github.com/gists/starred","team_url":"https://api.github.com/teams","user_url":"https://api.github.com/users/{user}","user_organizations_url":"https://api.github.com/user/orgs","user_repositories_url":"https://api.github.com/users/{user}/repos{?type,page,per_page,sort}","user_search_url":"https://api.github.com/search/users?q={query}{&page,per_page,sort,order}"}
msf-git (S:0 J:0)>
```

Check and save the egress IP address via Dyn-DNS to a file with -o

```
msf > request "http://checkip.dyndns.com/" -o my_ip.txt
[*] Wrote 106 bytes to /home/steiner/repos/msf-git/my_ip.txt
msf > cat my_ip.txt
[*] exec: cat my_ip.txt

<html><head><title>Current IP Check</title></head><body>Current IP Address: xx.xx.xx.xx</body></html>
msf-git (S:1 J:0) exploit(handler) >
```

Steps:
- [ ] Load the requests plugin
- [ ] View the help output to show option flags
- [ ] Pick an HTTP end point and make a GET request to it with parameters
- [ ] Make a POST request to it with parameters in -d
- [ ] Try different methods (OPTIONS and FAKE are fun)

### CODE REVIEW

**Comment by bcook-r7** [comment]
Nice, this looks very useful!

**Comment by hmoore-r7** [comment]
Might rename this from 'requests' to 'http_requests', or since is targeting curl, a riff on that (```curlish```?)

**Comment by kernelsmith** [comment]
Very cool. "Curling"?  "Curler"?
Tho, if u leave it as "request" it leaves the door open for other protocols to be added at a later date (after some refactoring of course). Usage could be msf > request type [opts] where opts are specific to the request type, so for type http, it would have the options you already included.

Just a thought

**Comment by zeroSteiner** [comment]
@kernelsmith that was my original thought so FTP could potentially be added. I've already pushed a change to rename the plugin but  I could change it back or future protocols could be implemented separately as proto_requests.

**Comment by kernelsmith** [comment]
Don't forget gopher ;). Yeah I was worried the proto_request naming convention would start to pollute the command name space if too many protos got added. I'm very keen to check out this code!

**Comment by wvu-r7** [comment]
Pretty cool, @zeroSteiner!

**Comment by wvu-r7** [comment]
Anyone want to pick this up? I think we can move forward with this as ```http_request```. If I really want FTP (or DICT, FILE, FTPS, GOPHER, HTTP, HTTPS, IMAP, IMAPS, LDAP, LDAPS, POP3, POP3S, RTMP, RTSP, SCP, SFTP, SMTP, SMTPS, TELNET, or TFTP), I can just shell out to ```curl```, which is what I'd normally do... Thoughts?

**Comment by hmoore-r7** [comment]
I find ```http_request``` to be kind of klunky, what about just ```http``` ?

**Comment by wvu-r7** [comment]
I like that! No conflict, so it works. I'd actually use that.

**Comment by kernelsmith** [comment]
You can't really shell out in these instances because you want the request to come from the session host, which is unlikely to have things like curl if it's windows. 

@zerosteiner and I have discussed a lil bit and are fond of the request $type syntax which allows for further expansion w/o polluting the namespace if anyone ever adds to the types. I could take up the PR if that's something agreeable. We're not married to that concept, it just seemed the most forward thinking

**Comment by wvu-r7** [comment]
@kernelsmith: Doh!

**Comment by wvu-r7** [comment]
:)

**Comment by bcook-r7** [comment]
I like it @kernelsmith. Go for it.

**Comment by kernelsmith** [comment]
cool, will do

**Comment by wvu-r7** [comment]
Thanks, @kernelsmith!! You're the best. :)

**Comment by kernelsmith** [comment]
@zeroSteiner I've made the mods.  Once I test, I will probably PR it back to you since it's a change in behavior

**Comment by kernelsmith** [comment]
I refactored this and have been in collab w/@zeroSteiner.  Giving it some testing now

**Comment by wvu-r7** [comment]
@kernelsmith: Rockin'. Thanks!

**Comment by bcook-r7** [comment]
Should we open a new PR when this is ready?

**Comment by kernelsmith** [comment]
I have a PR for @zeroSteiner but I haven't had a chance to send it

**Comment by kernelsmith** [comment]
PR sent and commented on and updated

**Comment by wvu-r7** [comment]
PR is zeroSteiner#11.

**Comment by zeroSteiner** [comment]
I landed the changes from @kernelsmith for this. One thing to note for testing is that all of the examples I gave in the original PR text need to be adjusted in the following ways:
 * The command is now simply ```request``` not ```http_request```
 * The URI needs to be the last argument not the first

**Comment by kernelsmith** [comment]
I updated the PR description to match the new syntax.  checking on the Travis failure now

**Comment by kernelsmith** [comment]
@zeroSteiner I think the PR needs to be rebased, I think that's what's causing the failures

**Comment by kernelsmith** [comment]
ok apparently I didn't save the changes to the description, updating again

**Comment by zeroSteiner** [comment]
Good call @kernelsmith, rebasing seems to have fixed the Travis issue.

**Comment by kernelsmith** [comment]
wow, even I had doubts about that ;)  But buster and the guys have been messing w/Travis's versions so...

File: plugins/request.rb [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+          options[:method] ||= 'POST'
+        when '-G'
+          options[:method] = 'GET'
+        when HELP_REGEX
+          #help(opt_parser)
+          # guard to prevent further option processing & stymie request handling
+          return [nil, opt_parser]
+        when '-H'
+          name, value = val.split(':', 2)
+          options[:headers][name] = value.strip
```
Comment (wvu-r7) [inline plugins/request.rb]: Maybe check for ```nil``` here? A ```to_s``` would work.

Comment (kernelsmith) [inline plugins/request.rb]: options[:headers][name] = value.to_s.strip if name

might work, tho a caller would have to handle if value.to_s.strip becomes nil

**Comment by wvu-r7** [comment]
Thank you, gents. :)

**Pull request merged**